### ISSUE TITLE
TRYING TO PORT UGTRAIN TO ANDROID :)

### ISSUE THREAD

**Issue by botbahlul** [ugtrain/ugtrain#82]
Hi, I just tried to port this to Thunder OS Lite v4.5 (a Phoenix OS moded distro, aka Android x86 moded) via Termux because previously I have succesfuly ported scanmem to this OS

at first it complained about missing execinfo.h
so I just picked up execinfo.h from libc6-dev debian package and put that on termux include folder /data/data/com.termux/files/usr/include

so I tried to build it again

but then next errors came

In file included from libmemdisc.c:34:
/data/data/com.termux/files/usr/include/execinfo.h:27:51: error: expected function body after function declarator
extern int backtrace (void **__array, int __size) __nonnull ((1));
                                                  ^
/data/data/com.termux/files/usr/include/execinfo.h:33:6: error: expected function body after function declarator
     __THROW __nonnull ((1));
     ^
/data/data/com.termux/files/usr/include/execinfo.h:39:6: error: expected function body after function declarator
     __THROW __nonnull ((1));
     ^
In file included from libmemdisc.c:36:
../../src/lib/system.h:82:10: warning: implicit declaration of function 'secure_getenv' is invalid in C99 [-Wimplicit-function-declaration]
        *home = secure_getenv("HOME");
                ^
../../src/lib/system.h:82:8: warning: incompatible integer to pointer conversion assigning to 'char *' from 'int' [-Wint-conversion]
        *home = secure_getenv("HOME");
              ^ ~~~~~~~~~~~~~~~~~~~~~
libmemdisc.c:285:13: warning: implicit declaration of function 'secure_getenv' is invalid in C99 [-Wimplicit-function-declaration]
        expected = secure_getenv(UGT_GAME_PROC_NAME);
                   ^
libmemdisc.c:285:11: warning: incompatible integer to pointer conversion assigning to 'char *' from 'int' [-Wint-conversion]
        expected = secure_getenv(UGT_GAME_PROC_NAME);
                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libmemdisc.c:297:2: warning: implicit declaration of function 'sigignore' is invalid in C99 [-Wimplicit-function-declaration]
        sigignore(SIGPIPE);
        ^
libmemdisc.c:655:14: warning: implicit declaration of function 'backtrace' is invalid in C99 [-Wimplicit-function-declaration]
        num_taddr = backtrace(trace, MAX_GNUBT);
                    ^
6 warnings and 3 errors generated.
make[3]: *** [Makefile:655: libmemdisc64_la-libmemdisc.lo] Error 1
make[3]: Leaving directory '/data/data/com.termux/files/home/ugtrain/src/linuxhooking'
make[2]: *** [Makefile:702: all-recursive] Error 1
make[2]: Leaving directory '/data/data/com.termux/files/home/ugtrain/src'
make[1]: *** [Makefile:463: all] Error 2
make[1]: Leaving directory '/data/data/com.termux/files/home/ugtrain/src'

so, my question is..

WHY DID IT GIVE ERROR WHEN COMPILE ON THUNDER OS BUT NO ERRORS WHEN COMPILE ON DEBIAN 9 (DEBIAN STRETCH)?

**Comment by sriemer** [comment]
Obviously, there is no `glibc`. Might there be a [bionic libc](https://en.wikipedia.org/wiki/Bionic_(software)) instead?

GNU `backtrace()` is crashing crap any way. I just added it for testing and comparing to the instruction pointer backwards search on the stack. You can drop that functionality for your port.

Also GNU `secure_getenv()` is not available of cause. You can replace it with `getenv()`. I might have implemented an inline function `util_getenv()` to wrap that so that only one location has to be touched.

Then only the `sigignore()` issue is remaining. Your libc violates POSIX. You could replace it with `sigaction()`.

Do you feel confident to do those changes or do you need further help with that? TIA

**Comment by sriemer** [comment]
Have you checked if `LD_PRELOAD` is usable on that platform at all or if security settings block it?

I once ported `ugtrain` to a Nokia N9 smart phone based on a Debian like distro (MeeGo 1.2 Harmattan). I had to hack the package manager so that I could request the required capabilities for `ptrace()` from the security module and `LD_PRELOAD` was not available. So I used a modified `patchelf` with ARM support and patched the lib dependency right into the executable. But after some time, some background security task also noticed the file modification and I always had to copy files first before `ugtrain` was running. But it worked with the most popular game on that platform. :smiley: 

I'm just saying be prepared for further issues on smart phones due to security measures.
